torch/_dynamo/variables/torch.py

```diff
@@ -911,6 +911,53 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
                 comparable_state,
             )
 
+        # See NOTE [HigherOrderOperator tracing design] for more details.
+        def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
+            try:
+                with tx.output.new_subtracer() as tracer:
+                    args = []
+                    # One argument to graph per sub_args
+                    for a in sub_args:
+                        if isinstance(a, TensorVariable):
+                            tracer.create_graph_input(a.as_proxy().node.name)
+                            args.append(a)
+                        else:
+                            # call_function() needs a TensorVariable, therefore we construct
+                            # one with inner graph proxy.
+                            assert isinstance(a, torch.Tensor)
+                            proxy = tracer.create_graph_input("arg")
+                            args.append(
+                                wrap_fx_proxy(tx=tx, proxy=proxy, example_value=a)
+                            )
+
+                    output = f.call_function(tx, args, {})
+                    # Register output to graph
+                    # Modeled off of compile_and_call_fx_graph
+                    # TODO: support non single Tensor output
+                    assert isinstance(output, TensorVariable)
+
+                    tx.output.guards.update(output.guards)
+                    tx.output.create_node(
+                        "output",
+                        "output",
+                        (tracer.create_arg((output.as_proxy(),))),
+                        {},
+                    )
+
+                    graph = tx.output.graph
+                    lifted_freevars = tracer.lifted_freevars
+
+                    return (
+                        output,
+                        graph,
+                        lifted_freevars,
+                    )
+
+            except torch._dynamo.exc.Unsupported as ex:
+                tx.output.graph = graph_checkpoint
+                tx.restore_graphstate(checkpoint)
+                raise
```

Comment:
This does give us two restores, is that a problem? If introspection of the body function raises Unsupported, then:
- we will restore_graphstate here
- we will restore_graphstate to the call to HigherOrderOperator

At any rate, I think we need to restore tx.output.graph because we may have mutated it if we saw any untracked free variables